prime-router/src/main/kotlin/fhirengine/engine/FHIRReceiverFilter.kt

```diff
@@ -28,24 +56,397 @@ class FHIRReceiverFilter(
+        override val message: String =
+            "Item was not routed to $receiverName because it did not pass the $filterType. Item failed on: $filter"
+
+        override val errorCode: ErrorCode = ErrorCode.UNKNOWN
     }
 
-    override val finishedField: Field<OffsetDateTime> = Tables.TASK.RECEIVER_FILTERED_AT
-    override val engineType: String = "ReceiverFilter"
+    private fun getBundleToRoute(receiver: Receiver, bundle: Bundle, actionLogger: ActionLogger): Bundle? {
+        val trackingId = bundle.identifier.value
+
+        if (doesBundlePassFilters(
+                receiver,
+                bundle,
+                actionLogger,
+                trackingId,
+                receiver.qualityFilter,
+                ReportStreamFilterType.QUALITY_FILTER
+            )
+        ) {
+            return null
+        }
+
+        if (doesBundlePassFilters(
+                receiver,
+                bundle,
+                actionLogger,
+                trackingId,
+                receiver.routingFilter,
+                ReportStreamFilterType.ROUTING_FILTER
+            )
+        ) {
+            return null
+        }
+
+        if (doesBundlePassFilters(
+                receiver,
+                bundle,
+                actionLogger,
+                trackingId,
+                receiver.processingModeFilter,
+                ReportStreamFilterType.PROCESSING_MODE_FILTER
+            )
+        ) {
+            return null
+        }
+
+        val prunedBundle = evaluateObservationConditionFilters(receiver, bundle, actionLogger, trackingId)
+
+        return prunedBundle
+    }
+
+    class MisconfiguredReceiverConditionFilters(val receiver: Receiver) : RuntimeException() {
+        override val message: String =
+            """${receiver.fullName} has \"conditionFilter\" and \"mappedConditionFilter\" which is not allowed as it can
+                | result in unintended bugs.  Please update the receiver to only use one
+""".trimMargin()
+    }
+
+    private fun evaluateObservationConditionFilters(
+        receiver: Receiver,
+        bundle: Bundle,
+        actionLogger: ActionLogger,
+        trackingId: String,
+    ): Bundle? {
+        val conditionFilters = receiver.conditionFilter
+        val mappedConditionFilters = receiver.mappedConditionFilter
+
+        // These filters are not compatible with each other and configuring them at the same time is likely to produce
+        // unintended results
+        if (conditionFilters.isNotEmpty() && mappedConditionFilters.isNotEmpty()) {
+            throw MisconfiguredReceiverConditionFilters(receiver)
+        }
+
+        val allObservations = bundle.getObservations()
+        val prunedBundle = if (conditionFilters.isNotEmpty()) {
+            val (keptObservations, filteredObservations) = allObservations.partition { observation ->
+                conditionFilters.any { filter ->
+                    FhirPathUtils.evaluateCondition(
```

Comment:
After some discussion we decided this should fail loudly. Previously it could not fail loudly because that would end routing early, but now it can fail since the side effect is limited to one receiver and the step is easy to reconstruct so to speak.